src/conv-core/converse.h

```diff
@@ -1098,21 +1098,33 @@ void  CmiError(const char *format, ...);
 
 #define __CMK_XSTRING(x) __CMK_STRING(x)
 
-extern void __cmi_assert(const char *);
+// These wrappers of CmiAbort are necessary to prevent the stringified
+// expression from being treated as a format string (if it contains a
+// %, for example)
+extern void __cmi_assert(const char*);
+#if defined __GNUC__ || defined __clang__
+__attribute__((format(printf, 2, 3)))
+#endif
+extern void __cmi_assert_msg(const char* const, const char* const, ...);
+
 #define CmiEnforce(expr) \
   ((void) ((expr) ? 0 :                   \
      (__cmi_assert ("Assertion \"" __CMK_STRING(expr) \
                     "\" failed in file " __FILE__ \
                     " line " __CMK_XSTRING(__LINE__) "."), 0)))
-
-#if ! CMK_ERROR_CHECKING
-#define CmiAssert(expr) ((void) 0)
+#define CmiEnforceMsg(expr, msg, ...) \
+  ((void) ((expr) ? 0 : \
+     (__cmi_assert_msg ("Assertion \"" __CMK_STRING(expr) \
+                        "\" failed in file " __FILE__ \
+                        " line " __CMK_XSTRING(__LINE__) ".", \
+                        msg, __VA_ARGS__), 0)))
```

Comment:
Could you pass the whole thing to `CmiAbort`?

Reply:
No, as the comment on line 1101 says, the issue there is that the stringified `expr` may have a `%` in it (e.g. `var%d == 0` to assert that `var` is a multiple of `d`), which can then cause problems when it is treated as a format string. Calling `CmiAbort` through this `__cmi_assert_msg` wrapper lets us handle the macro generated and user provided strings separately.

Reply:
Ah, I see, I think your edited version makes sense. I'll update this branch with something of that ilk.

Reply:
Added in a0d0933.

Reply:
Actually, it seems the answer is no, calling `CmiAbort` directly is an issue because it can be called in any scope and there are other implementations of `CmiAbort`, namely in `CkArray` and `CkMigratable`, etc. This causes issues when there are asserts from a static context because `CmiAbort` resolves to the version in the inheritance tree, which then fails because it can't be called statically, so we need to call it through some kind of wrapper to make sure that this always resolves correctly.

Reply:
Also, I think it's worthwhile to create a function to pass the user message into the `CmiAbort` to ensure that the user message is printed at the end of the output; if we just print it out and then call abort, then it'll be above the stack trace and other printed errors, so it can be hard to find.